File: jetbrains-plugin/src/main/kotlin/com/mariadbprofiler/plugin/service/StatisticsService.kt
package com.mariadbprofiler.plugin.service

import com.intellij.openapi.components.Service
import com.intellij.openapi.project.Project
import com.mariadbprofiler.plugin.model.QueryEntry
import com.mariadbprofiler.plugin.model.QueryType

@Service(Service.Level.PROJECT)
class StatisticsService(private val project: Project) {

    data class QueryStats(
        val totalQueries: Int,
        val byType: Map<QueryType, Int>,
        val byTable: Map<String, Int>,
        val byTag: Map<String, Int>,
        val timelinePoints: List<TimelinePoint>
    )

    data class TimelinePoint(
        val timestamp: Double,
        val queryCount: Int,
        val queryType: QueryType
    )

    fun computeStats(entries: List<QueryEntry>): QueryStats {
        val byType = entries.groupBy { it.queryType }
            .mapValues { it.value.size }

        val byTable = mutableMapOf<String, Int>()
        entries.forEach { entry ->
            entry.tables.forEach { table ->
                byTable[table] = (byTable[table] ?: 0) + 1
            }
        }

        val byTag = mutableMapOf<String, Int>()
        entries.forEach { entry ->
            entry.tags.forEach { tag ->
                byTag[tag] = (byTag[tag] ?: 0) + 1
            }
        }

        val timeline = entries.mapIndexed { index, entry ->
            TimelinePoint(
                timestamp = entry.timestamp,
                queryCount = index + 1,
                queryType = entry.queryType
            )
        }

        return QueryStats(
            totalQueries = entries.size,
            byType = byType,
            byTable = byTable.toMap(),
            byTag = byTag.toMap(),
            timelinePoints = timeline
        )
    }
}

File: vscode-extension/src/service/StatisticsService.ts
import { QueryEntry, getQueryType, getTables } from '../model/QueryEntry';

export interface QueryStats {
  totalQueries: number;
  byType: Record<string, number>;
  byTable: Record<string, number>;
  byTag: Record<string, number>;
}

export class StatisticsService {
  computeStats(entries: QueryEntry[]): QueryStats {
    const byType: Record<string, number> = {};
    const byTable: Record<string, number> = {};
    const byTag: Record<string, number> = {};

    for (const entry of entries) {
      // By type
      const qtype = getQueryType(entry);
      byType[qtype] = (byType[qtype] || 0) + 1;

      // By table
      for (const table of getTables(entry)) {
        byTable[table] = (byTable[table] || 0) + 1;
      }

      // By tag
      if (entry.tag) {
        byTag[entry.tag] = (byTag[entry.tag] || 0) + 1;
      }
    }

    return {
      totalQueries: entries.length,
      byType: sortByValueDesc(byType),
      byTable: sortByValueDesc(byTable),
      byTag: sortByValueDesc(byTag),
    };
  }
}

function sortByValueDesc(record: Record<string, number>): Record<string, number> {
  const sorted: Record<string, number> = {};
  const entries = Object.entries(record).sort((a, b) => b[1] - a[1]);
  for (const [key, value] of entries) {
    sorted[key] = value;
  }
  return sorted;
}
